runtime/Go/antlr/case_changing_stream.go

```diff
@@ -0,0 +1,37 @@
+package antlr
+
+import (
+	"unicode"
+)
+
+// CaseChangingStream wraps an existing CharStream, but upper cases, or
+// lower cases the input before it is tokenized.
+type CaseChangingStream struct {
+	CharStream
+
+	upper bool
+}
+
+// NewCaseChangingStream returns a new CaseChangingStream that forces
+// all tokens read from the underlying stream to be either upper case
+// or lower case based on the upper argument.
+func NewCaseChangingStream(in CharStream, upper bool) *CaseChangingStream {
+	return &CaseChangingStream{
+		in, upper,
+	}
+}
+
+// LA gets the value of the symbol at offset from the current position
+// from the underlying CharStream and converts it to either upper case
+// or lower case.
+func (is *CaseChangingStream) LA(offset int) int {
+	in := is.CharStream.LA(offset)
+	if in < 0 {
+		// Such as antlr.TokenEOF which is -1
+		return in
+	}
+	if is.upper {
+		return int(unicode.ToUpper(rune(in)))
```

Comment:
Maybe it's better to convert case at the beginning one time and use upper/lower values from a prepared array for the performance reason? `LA` function can be called multiple times.

Reply:
I could cache the result, but I didn't want this to work by reading the full input up front, because this data could be streamed, or extremely large.

However, I don't think caching will save much, as this is quite a cheap operation for the CPU. But I'm happy to benchmark it.

Reply:
Ok, got it. Maybe it's right.